### ISSUE TITLE
Definition request: lodash/fp

### ISSUE THREAD

**Issue by willfrew** [DefinitelyTyped/DefinitelyTyped#7896]
In Lodash 4.X, they introduced core support for a more functional version of the library.
By importing [`'lodash/fp'`](https://github.com/lodash/lodash/tree/master/fp), you get access "to an immutable, auto-curried, iteratee-first, data-last version."

The conversion is done automatically, and basically just uses their own [`rearg`](https://lodash.com/docs#rearg) method to move around the arguments in a [standard way](https://github.com/lodash/lodash/blob/master/fp/_mapping.js#L119)*.

I'm pretty new to writing Typescript defintions but I hope that we wouldn't need a full-blown, modified version of the existing [`lodash.d.ts`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/lodash/lodash.d.ts) and that we could just describe the conversion function and 'transform' the existing definition, if that's even possible.

\* It looks like there are a few [exceptions](https://github.com/lodash/lodash/blob/master/fp/_mapping.js#L133), but I don't think they'd be too problematic.

**Comment by RIAstar** [comment]
I'm very interested in this as well. 
However, as much as I would like your 'transforming' idea to work, I'm afraid it won't do: it's not merely reordering the arguments, but also making every function curried; in many cases some arguments are dropped (because we need a fixed arity) and optional arguments become non-optional. Moreover each argument of a curried function can potentially be a placeholder.

Then again, I might be wrong. And I sure hope I am: I've been writing 'lodah/fp' typedefs as I required them and I can tell you it's a royal pain in the butt. TypeScript just doesn't have good support for FP paradigms. More specifically, it lacks a way of describing a curried function on one line. With such definitions it is extremely easy to "loose" a type somewhere along the way :(

An example definition for `filter`:
```
    interface filter {
        // filter list
        <T> (predicate:string | ListIterator <T, boolean>): <T> (collection:List <T>) => T[];
        <T> (predicate:string | ListIterator <T, boolean>, collection:List <T>): T[];

        // filter dictionary
        <T> (predicate:string | DictionaryIterator <T, boolean>): <T> (collection:Dictionary <T>) => T[];
        <T> (predicate:string | DictionaryIterator <T, boolean>, collection:Dictionary <T>): T[];

        // filter string
        (predicate:StringIterator <boolean>): (collection:string) => string[];
        (predicate:StringIterator <boolean>, collection:string): string[];

        // placeholders
        <T> (predicate:string | ListIterator <T, boolean> | DictionaryIterator <T, boolean>, collection:PH): <T> (collection:List <T> | Dictionary <T>) => T[];
        <T> (predicate:PH, collection:List <T> | Dictionary <T>): <T> (predicate:string | ListIterator <T, boolean> | DictionaryIterator <T, boolean>) => T[];
        (predicate:StringIterator <boolean>, collection:PH): (collection:string) => string[];
        (predicate:PH, collection:string): (predicate:StringIterator <boolean>) => string[];
        <O extends {}, T> (predicate:O, PH): <O extends {}, T> (predicate:O) => T;

        // object predicate (must come after placeholders, because they're objects too)
        <O extends {}, T> (predicate:O): <T> (collection:List <T> | Dictionary <T>) => T[];
        <O extends {}, T> (predicate:O, collection:List <T> | Dictionary <T>): T[];
    }
```

This is a function with arity 2. The complexity increases exponentially with the arity...
You can drop all the definitions for Lodash's implicit and explicit wrappers though.

**Comment by artaommahe** [comment]
Maybe smhing about lodash-fp to definitons? can't use with typescript
```js
import * as lodash from 'lodash/fp'
```
because of 
```
Error TS2307: Cannot find module 'lodash/fp'.
```

**Comment by mlegenhausen** [comment]
+1

**Comment by jonaskello** [comment]
I guess this will be similar to the work being done for [rambda definitions](https://github.com/donnut/typescript-ramda). They seem to restrict it to applying functions in two steps which might be a way to deal with the complexity of functions with arity higer than 2.

**Comment by Yord** [comment]
+1

**Comment by vvakame** [comment]
IMO, First, it is preferable to pass through the compilation.

like this.

```
declare module "lodash/fp" {
  import _ = require("lodash");
  export = _;
}
```

**Comment by stepancar** [comment]
@artaommahe @artaommahe  @willfrew  check this [PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/9052)
 please

**Comment by stepancar** [comment]
@willfrew, i think this issue should be closed

**Comment by MichalZalecki** [comment]
@stepancar Do you know why your typings are gone? Function export doesn't work any more.

**Comment by jmlopez-rod** [comment]
Currently the type definitions for lodash has this

```
declare module "lodash/fp" {
    export = _;
}
```

Is does not seem to work properly. When I do

```
import * as fp from 'lodash/fp`;

...
fp.flow(fp.filter((x) => {})(items);
```

It complains about the argument to filter since that points to the filter function from lodash and not the fp version of it. I was planning on overwritting those definitions but now I can't do it since `node_modules/@types/lodash` already declares it and it doesn't let me redeclare it. Any solutions for this?

**Comment by willfrew** [comment]
Sorry for the unresponsiveness on this - but the PR does **not** solve the issue with `lodash/fp`. Order of arguments is systematically transformed for most lodash functions for fp so just re-exporting as `lodash/fp` is totally incorrect.

I suggest the re-export is removed until a proper solution can be created.

**Comment by jonaskello** [comment]
The typings will probably be easier to do once typescript gets [Variadic Kinds](https://github.com/Microsoft/TypeScript/issues/5453). This function is currently on the roadmap for typescript 2.1.

**Comment by jmlopez-rod** [comment]
So... reopen issue?

**Comment by jmlopez-rod** [comment]
Can we go ahead and remove this for the time being?

```
declare module "lodash/fp" {
    export = _;
}
```

For now make users who are using fp to write their own temporary definition.

**Comment by nojaf** [comment]
An example of a temporary definition file:

>lodash-fp.d.ts
```typescript
declare module 'lodash/fp' {
    export function startsWith(): boolean;
    export function contains(): boolean;
    export function lowerCase(a: string): string;
    export function map(f: Function, i: any): any;
    export function uniq(a: any[]): any[];
}
```
I add what I need along the way just so TypeScript doesn't complain anymore.

**Comment by albohlabs** [comment]
[I've done this too.](https://gist.github.com/albohlabs/1c864d83b17e4e99edff210729327e03)

**Comment by nahuel** [comment]
why not integrating @albohlabs typings? seems like a good start.

**Comment by amir-arad** [comment]
would love that

**Comment by spoonscen** [comment]
@amir-arad Can you submit a PR?

**Comment by wildeyes** [comment]
Any typed libraries that provides similar functionality?

**Comment by Ebuall** [comment]
@wildeyes you can try ramda, it is typed, but not perfect. There is also a new library designed specifically for typescript https://github.com/TylorS/typed

**Comment by FranzSkuffka** [comment]
Also enhanced it a bit - noImplicitAny will now not be triggered because _.map infers the type of the function.
https://gist.github.com/FranzSkuffka/d9c52982c265d44a1093f218664b6a5f/revisions

**Comment by FranzSkuffka** [comment]
Is there some PR already?

**Comment by amir-arad** [comment]
sadly, no. I dont have the capacity to make a proper PR for this project. I'm still using [this](https://gist.github.com/amir-arad/27adf36e33a06c5d65f19aa4bab272dc) inside my project.

**Comment by amir-arad** [comment]
I bet it breaks spectacularly on 2.6 strict mode though.

**Comment by FranzSkuffka** [comment]
@amir-arad - We're at 2.6.2 and it works fine for us. Do you expect it to break on any compile?

![image](https://user-images.githubusercontent.com/7485959/34099177-0b248d96-e3df-11e7-895b-c3ba17930101.png)

**Comment by amir-arad** [comment]
I've had issues with other functional APIs using TS 2.6. guess this one is ok.

**Comment by aj-r** [comment]
Unfortunately, variadic kinds still aren't implemented in 2.7, and I wouldn't count on them until at least 2.9. The best thing for now is probably to write a script that generates the types. I'm experimenting with using `lodash/fp/convert` itself to help generate these types, with some success so far.
